### PR TITLE
ClearFaults Action

### PR DESCRIPTION
**Pull request by AstroStucky**
## Linked Issues:
Jira Ticket 🎟️ | [OCEANWATER-1198](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-1198) |

## Summary of Changes
* Refactoring of how action servers interface with faults. 
  * Shortened name of faults_interface.py to faults.py
* **Fault Behavior Changes**:
  * GOAL_ERROR faults in `/system_faults_status` will no longer automatically clear when the next action in the group is called.
  * **FaultClear** action has been added to facilitate clearing of only GOAL_ERROR faults.
  * Subsystem faults (e.g. `/arm_faults_status`) are cleared at the start of the next action called in that subsystem.
    * Presently, only TRAJECTORY_GENERATION in `/arm_faults_status` has been implemented, and the other subsystem messages (`/camera_faults_status` and `/pan_tilt_fault_status`) never change. 
  * An action called within a group that is faulted will be rejected. While the log message use the language "rejected", to actionlib the action is merely ABORTED since we use SimpleActionServer at present. 

## Test (quick, non-comprehenisve)
This test will be thorough about all possible fault behaviors with only a single action. The later test will be comprehensive of all actions. All reviewers, please do this test. 
1. Launch any world.
2. In 2nd terminal `rostopic echo /system_faults_status`
3. In 3rd terminal `rostopic echo /arm_faults_status`
4. Call an invalid ArmMoveCartesian `rosrun ow_lander arm_move_cartesian.py`. 
5. Wait a little. Simulator output should be (nothing's changed since previous versions)
```
[INFO:/lander_action_servers] ArmMoveCartesian action started
[ERROR:/move_group] arm/arm: Unable to sample any valid states for goal tree
[ WARN:/move_group] Timed out
[ WARN:/move_group_commander_wrappers_1693511951616356524] Fail: ABORTED: TIMED_OUT
[ERROR:/lander_action_servers] ArmMoveCartesian: Aborted - MoveIt planning failed with error code: val: -6                                                                                    
[INFO:/lander_action_servers] ArmMoveCartesian action complete
```
6. Both your `/system_faults_status` and `/arm_faults_status` output should show a value of 2. The equality of these values is coincidental: one indicates an ARM_GOAL_ERROR and the other indicates a TRAJECTORY_GENERATION fault.
7. Call another ARM group action `rosrun ow_lander arm_unstow.py`. Immediately simulator output should be 
```
[INFO:/lander_action_servers] ArmMoveCartesian action started
[ERROR:/lander_action_servers] ArmMoveCartesian: Rejected - Action group, ARM, is faulted. Use the  FaultClear action to make the group operable again.                                       
[INFO:/lander_action_servers] ArmMoveCartesian action complete
```
8. Both your `/system_faults_status` and `/arm_faults_status` output should both still be a value of 2. 
9. Call the FaultClear action to clear the ARM_GOAL_ERROR (a bitmask value of 2) `rosrun ow_lander fault_clear.py 2`
10. `/system_faults_status` should show 0, but `/arm_faults_status` will still show 2. 
11. Call a valid ArmMoveCartesian `rosrun ow_lander arm_move_cartesian.py -x 1.8 -z 0.5`. 
12. Both `/system_faults_status` and `/arm_faults_status` should now be 0 because the arm subsytem faults are reset at the beginning of the action. 
13. The action should complete without trouble, and by the end of it both fault messages will still be 0. 

## Test (comprehensive of every affected action)
We really should test this feature for every single action it affects before we call it working, but this takes a decent amount of time, so I will split it among both myself and the reviewers. Not all actions need to be tested because they either never fail, are too difficult to cause a failure, or do not have a fault_handler assigned to them. 

The procedure, generalized to any action, is to
1. Call the action with invalid parameters
2. Check that the value of `/system_fault_status` is the appropriate GOAL_ERROR
3. Call FaultClear providing the value of `/system_fault_status`.
4. `/system_fault_status` should now be 0.
5. Call the action with valid parameters, and ensure it works.  

### Test all ARM Group Actions
@Hao12450 I am assigning this to you. 
**Value of ARM_GOAL_ERROR:** 2
**Method of Invalidation:** Command poses that are out of reach to the arm, or invalid joint angles. 
Test the following actions using the above procedure:
* ArmMoveJoint
* ArmMoveJoints
* ArmMoveJointsGuarded
* ~~ArmMoveCartesian~~ Already tested by previous PR test
* ArmMoveCartesianGuarded
* ArmFindSurface

### Test all TASK group actions
@kmdalal I am assigning this to you. Test the following actions using above procedure. 
**Value of TASK_GOAL_ERROR:** 8
**Method of Invalidation:** Command dig points or discard points that are well outside of the workspace (e.g. x = 100)
Test the following actions using the above procedure:
* TaskGrind - Note that task_grind.py uses positional arguments unlike the other clients. An example of invalid input is `rosrun ow_lander task_grind.py 100`
* ~TaskDeliverSample~ No easy way to fault this, so skip.
* TaskDiscardSample
* TaskScoopCircular
* TaskScoopLinear

### Test PAN_TILT and CAMERA group actions
These I will leave to myself, I will include the necessary details here in case reviewers would like to test them too. 
**Value of CAMERA_GOAL_ERROR:** 16
**Method of Invalidation:** Tick `camera_left_trigger_failure` to true in rqt_gui
* CameraCapture
* ~~CameraSetExposure`~ - Only updates a topic, don't think there is a way to fail it.

And for the PAN_TILT actions...
**Value of CAMERA_GOAL_ERROR:** 64
**Method of Invalidation:** Command an impossible tilt angle. For the pan-only action, inject a locked joint fault.
* PanTiltMoveJoints
* PanTiltMoveCartesian
* Pan
* Tilt

### CODE REVIEW

**Comment by AstroStucky** [comment]
Completed **Test PAN_TILT and CAMERA group actions** and it passed

**Comment by Hao12450** [comment]
All tests passed!

File: ow_lander/src/ow_lander/faults.py [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+  def notify_preempted(self):
+    self._notify(GoalStatus.PREEMPTED)
+
+  def is_faulted(self):
+    return self._goal_fault_publisher.get_group_status(
+      self._group_id) == GoalStatus.ABORTED
+
+  def get_rejected_message(self):
+    group = self._goal_fault_publisher.get_group_str(self._group_id)
+    return f"Action group, {group}, is faulted. Use the  FaultClear action " \
```
Comment (Hao12450) [inline ow_lander/src/ow_lander/faults.py:105]:
```suggestion
    return f"Action group, {group}, is faulted. Use the FaultClear action " \
```
Will that better leave only one space between "the" and "Faultclear"?

**Comment by kmdalal** [comment]
My assigned tests all passed.   I failed them all the same way, by using x=100.

Note that all the prescribed tests involve inducing planning errors.  A more thorough test would include execution-time goal errors, but it seems hard to induce these.

I'm starting on regression tests using Plexil.

**Comment by kmdalal** [comment]
ReferenceMission1 produced goal errors (as has been recently happening) and this botched the run in a bad way -- the plan plowed through until finished, generating a stream of error messages about attempting operations while there's an uncleared goal error.

The PLEXIL interface to FaultClear is a straightforward task, yet still needs doing.  The tougher task is figuring out how and where to use it across the current set of plans.  This is important work that should keep going, but my inclination is to not squeeze it into Release 12 at this stage...

**Comment by kmdalal** [comment]
On reflection, this work strikes me as best an Epic with three tickets:

1. This ticket, which I think is almost ready for merge.
2. PLEXIL interface to FaultClear, which could be done within a day.
3. Augment the plan set to deal with goal errors, which I think is a bigger task needing some discussions and perhaps a lot of plan edits.

**Comment by AstroStucky** [comment]
> This is important work that should keep going, but my inclination is to not squeeze it into Release 12 at this stage...

Agreed.

I'll write the epic and the tickets, and start rebasing this branch onto an epic branch.

**Comment by AstroStucky** [comment]
For Release 12, do we want to stick with the goal error behavior as it is now in noetic-devel: goal errors will not clear/refault until the end of the next called action in the group. 

Or do we want to cherry-pick the slightly different behavior that I built with the very first commit of this branch: goal errors clear at the beginning of the next called action in the group and may be refaulted before the end of it. 

I could easily cherry-pick the commit into a new branch, and write up a very quick PR for it.

**Comment by kmdalal** [comment]
For Release 12, clearing the goal error at the start of the next action would be a good improvement.  I could review it promptly.

File: ow_lander/src/ow_lander/faults.py (hunk 1 above)
Comment (AstroStucky) [reply ow_lander/src/ow_lander/faults.py:105]: Thank you!

**Pull request closed without merging**